R/model-tree.R

```diff
@@ -95,74 +108,90 @@ model_tree <- function(
     .log_df,
     include_info = c("description","star", "tags"),
     color_by = "run",
+    size_by = NULL,
     add_summary = TRUE,
     digits = 3,
     zoomable = FALSE,
     static = FALSE,
     width = NULL,
     height = NULL,
+    font_size = 10,
     ...
 ){
   UseMethod("model_tree")
 }
 
-#' @rdname model_tree
+
 #' @export
 model_tree.character <- function(
     .log_df,
     include_info = c("description","star", "tags"),
     color_by = "run",
+    size_by = NULL,
     add_summary = TRUE,
     digits = 3,
     zoomable = FALSE,
     static = FALSE,
     width = NULL,
     height = NULL,
+    font_size = 10,
     ...
 ){
   checkmate::assert_directory_exists(.log_df)
   .log_df <- run_log(.log_df, ...)
   model_tree(
     .log_df,
-    include_info = include_info, color_by = color_by,
+    include_info = include_info,
+    color_by = color_by, size_by = size_by,
     add_summary = add_summary, digits = digits,
     zoomable = zoomable, static = static,
-    width = width, height = height
+    width = width, height = height,
+    font_size = font_size
   )
 }
 
-#' @rdname model_tree
 #' @export
 model_tree.bbi_log_df <- function(
     .log_df,
-    include_info = c("description","star", "tags"),
+    include_info = c("description", "star", "tags"),
     color_by = "run",
+    size_by = NULL,
     add_summary = TRUE,
     digits = 3,
     zoomable = FALSE,
     static = FALSE,
     width = NULL,
     height = NULL,
+    font_size = 10,
     ...
 ){
   # Make sure required dependencies are installed
   stop_if_tree_missing_deps(static = static)
 
   # Make tree data
-  tree_data <- make_tree_data(.log_df, include_info, color_by, add_summary)
+  # TODO: decide if we want to append color_by and size_by to tooltip automatically
+  # include_info <- unique(c(include_info, color_by, size_by))
+  tree_data <- make_tree_data(.log_df, include_info, color_by, size_by, add_summary)
 
   # Format coloring
   tree_data <- color_tree_by(tree_data, color_by = color_by)
+  tree_attr <- ifelse(is.null(color_by), "leafCount", color_by)
+
+  # Format sizing
+  tree_data <- size_tree_by(tree_data, size_by = size_by)
+  node_size <- ifelse(is.null(size_by), "leafCount", "node_size")
 
   # Compile attributes into tooltip
-  tree_data <- make_tree_tooltip(tree_data, digits = digits)
+  tree_data <- make_tree_tooltip(tree_data, digits = digits, font_size = font_size)
 
   # Create model tree
-  tree_attr <- ifelse(is.null(color_by), "leafCount", color_by)
   pl_tree <- collapsibleTree::collapsibleTreeNetwork(
-    tree_data, zoomable = zoomable, attribute = tree_attr,
-    fill="col", collapsed = FALSE, nodeSize = "leafCount",
-    tooltipHtml = "tooltip", width = width, height = height
+    tree_data, zoomable = zoomable, collapsed = FALSE,
+    # Coloring and sizing
+    attribute = tree_attr, fill="col", nodeSize = node_size, aggFun = identity,
```

Comment:
Based on dropping this `aggfun` and testing, I think setting this to `identity` (or at least something other than the default value of `sum`) is necessary to get the desired sizing, but it's not obvious to me what's going on here. An aggregation function is a `[]T -> T` function, reducing many values to one (which aligns with the `sum` default value of `collapsibleTree`). How does `identity` (`[]T -> []T`) behave here given that it doesn't reduce?

I'm asking to

1.  get a better understanding of what's going on here

2.  make sure we're not relying on undefined behavior

3.  get a code comment added, since I think the behavior here is non-obvious

Reply:
@kyleam I did a lot of investigating about `aggFun` and left details in the comment made in [0cf42e4](https://github.com/metrumresearchgroup/bbr/pull/720/commits/0cf42e40fa2ece78a977f4ab5db14f7942479c04).

passing `aggFun = identity` is not the same as `aggFun = base::identity` because `collapsibleTree` has specific handling for when that name is matched (see [here](https://github.com/AdeelK93/collapsibleTree/blob/master/R/collapsibleTreeNetwork.R#L204-L220)). There is additional `data.tree` nuance too, but I couldnt find a way to fix the issue of inconsistent sizing. I tried providing a custom `aggFun` and changing the data beforehand but nothing seemed to work. Im ok to merge with this known "quirk", but let me know if this should be mentioned in the documentation too.

Reply:
Thanks for looking into it.  A comment there is definitely helpful.

> Im ok to merge with this known "quirk", but let me know if this should be mentioned in the documentation too.

I think it's fine to leave as is too, at least for now.